ntoskrnl/ps/psmgr.c

```diff
@@ -62,106 +62,31 @@ BOOLEAN PspDoingGiveBacks;
-              IN PUSHORT OrdinalTable)
-{
-    ULONG Mid;
-    LONG Ret;
-
-    /* Fail if no names */
-    if (!NumberOfNames) return -1;
-
-    /* Do binary search */
-    Mid = NumberOfNames >> 1;
-    Ret = strcmp(Name, (PCHAR)((ULONG_PTR)DllBase + NameTable[Mid]));
-
-    /* Check if we found it */
-    if (!Ret) return OrdinalTable[Mid];
-
-    /* We didn't. Check if we only had one name to check */
-    if (NumberOfNames == 1) return -1;
-
-    /* Check if we should look up or down */
-    if (Ret < 0)
-    {
-        /* Loop down */
-        NumberOfNames = Mid;
-    }
-    else
-    {
-        /* Look up, update tables */
-        NameTable = &NameTable[Mid + 1];
-        OrdinalTable = &OrdinalTable[Mid + 1];
-        NumberOfNames -= (Mid - 1);
-    }
-
-    /* Call us recursively */
-    return NameToOrdinal(Name, DllBase, NumberOfNames, NameTable, OrdinalTable);
-}
-
-CODE_SEG("INIT")
+/* See mm/ARM3/sysldr.c */
+static CODE_SEG("INIT")
 NTSTATUS
-NTAPI
-LookupEntryPoint(IN PVOID DllBase,
-                 IN PCHAR Name,
-                 OUT PVOID *EntryPoint)
+LookupEntryPoint(
+    _In_ PVOID DllBase,
+    _In_ PCSTR Name,
+    _Out_ PVOID* EntryPoint)
 {
-    PULONG NameTable;
-    PUSHORT OrdinalTable;
-    PIMAGE_EXPORT_DIRECTORY ExportDirectory;
-    ULONG ExportSize;
-    CHAR Buffer[64];
-    USHORT Ordinal;
-    PULONG ExportTable;
-
-    /* Get the export directory */
-    ExportDirectory = RtlImageDirectoryEntryToData(DllBase,
-                                                   TRUE,
-                                                   IMAGE_DIRECTORY_ENTRY_EXPORT,
-                                                   &ExportSize);
-
-    /* Validate the name and copy it */
-    if (strlen(Name) > sizeof(Buffer) - 2) return STATUS_INVALID_PARAMETER;
-    strcpy(Buffer, Name);
-
-    /* Setup name tables */
-    NameTable = (PULONG)((ULONG_PTR)DllBase +
-                         ExportDirectory->AddressOfNames);
-    OrdinalTable = (PUSHORT)((ULONG_PTR)DllBase +
-                             ExportDirectory->AddressOfNameOrdinals);
-
-    /* Get the ordinal */
-    Ordinal = NameToOrdinal(Buffer,
-                            DllBase,
-                            ExportDirectory->NumberOfNames,
-                            NameTable,
-                            OrdinalTable);
-
-    /* Make sure the ordinal is valid */
-    if (Ordinal >= ExportDirectory->NumberOfFunctions)
-    {
-        /* It's not, fail */
-        return STATUS_PROCEDURE_NOT_FOUND;
-    }
+    /* Call the internal function */
+    NTSTATUS Status = RtlpFindExportedRoutineByName(DllBase,
+                                                    Name,
+                                                    EntryPoint,
+                                                    FALSE);
+    /* Adjust the returned status if we failed */
+    if (!NT_SUCCESS(Status) && (Status != STATUS_INVALID_PARAMETER))
+        Status = STATUS_PROCEDURE_NOT_FOUND;
```

Comment:
Is this something that should be logged?

Reply:
I don't think so. Original code didn't log here in case the entry point was not found. (Not sure it is useful as well, since this function in psmgr, and the one that calls it, are used so far only when loading exports from ntdll.dll.)